src/Val.cc

```diff
@@ -1761,6 +1761,100 @@ static void find_nested_record_types(const TypePtr& t, std::set<RecordType*>* fo
 		}
 	}
 
+using PatternValPtr = IntrusivePtr<PatternVal>;
+
+// Support class for returning multiple values from a table[pattern]
+// when indexed with a string.
+class TablePatternMatcher
+	{
+public:
+	TablePatternMatcher(const TableVal* _tbl, TypePtr _yield) : tbl(_tbl)
+		{
+		vtype = make_intrusive<VectorType>(std::move(_yield));
+		}
+	~TablePatternMatcher() { Clear(); }
+
+	void Insert(ValPtr pat, ValPtr yield) { Clear(); }
+	void Remove(ValPtr pat) { Clear(); }
+
+	void Clear()
+		{
+		delete matcher;
+		matcher = nullptr;
```

Comment:
Opportunity for another unique_ptr and `reset()`? Wouldn't need a destructor then either.